### PR TITLE
507 GitHub login

### LINKED ISSUE #507: Add OAuth2 login functionality into cobra CLI
<a href="https://opensource.globo.com/hacktoberfest/"></a>

## Motivation

To have features like vulnerability management (and many others), we need to have some kind of authentication in huskyCI. 

If we have a CLI command that will log in users using Github and Gitlab, for instance, we will have the possibility to check which repositories a developer has access to, profile image, and many others. 

## It would be great if

We have an OAuth2 solution coded in our [CLI](https://github.com/globocom/huskyCI/tree/master/cli) by implementing a [cobra](https://github.com/spf13/cobra) command `login`.

## What we expect

By typing the following command, huskyCI must be able to redirect the authorization request to an OAuth2 provider (Github? ❤️) open a web browser for the user to complete the login, and store the token into a file `.huskyci` to be used later on.

```sh
$ huskyci login
Successfully logged in!
```

## Tips

* [Gsh already does something similar](https://github.com/globocom/gsh/blob/master/cli/cmd/login.go)  
* [Tsuru already does something similar as well](https://github.com/tsuru/tsuru/blob/master/cmd/oauth.go) 
* [Authorizing OAuth Apps - Github](https://developer.github.com/apps/building-oauth-apps/authorizing-oauth-apps/)

### PR DESCRIPTION
**Pull request by lzakharov**
### Description

Adds GitHub authorization command.

Closes #507

### Proposed Changes

You can log in to GitHub via `login` CLI command.

### Testing

```
make build-cli
./cli/huskyci-cli-bin login
cat .huskyci
```

### CODE REVIEW

File: cli/pkg/github/github.go [hunk 1]
```diff
@@ -0,0 +1,16 @@
+package github
+
+import (
+	"net/url"
+)
+
+const (
+	// ClientID is a GitHub App client id.
+	ClientID = "9c1379d55962c56e11a8"
```
Comment (lzakharov) [inline cli/pkg/github/github.go:9]: @rafaveira3 here should be another client id.

File: cli/cmd/login.go [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+	"github.com/pkg/browser"
+	"github.com/spf13/cobra"
+
+	"github.com/globocom/huskyCI/cli/pkg/github"
+)
+
+var loginCmd = &cobra.Command{
+	Use:   "login",
+	Short: "Log in to GitHub",
```
Comment (Krlier) [inline cli/cmd/login.go:19]: What do you think of changing this line to "Log in with GitHub"?

File: cli/pkg/github/auth.go [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+}
+
+// NewDeviceFlow creates a new GitHub device authorization flow.
+func NewDeviceFlow(baseURL *url.URL, client *http.Client) DeviceFlow {
+	return DeviceFlow{baseURI: baseURL, client: client}
+}
+
+// GetCodes requests the device and user verification codes from GitHub.
+func (df DeviceFlow) GetCodes(req *GetCodesRequest) (*GetCodesResponse, error) {
+	uri := df.uri("login/device/code")
```
Comment (Krlier) [inline cli/pkg/github/auth.go:26]: Would you mind adding this path to `cli/pkg/github/github.go` and use it as a constant as well? By doing this, if it changes, we'd need to simply change the value in `github.go`. Same as the `baseURL` variable.

File: cli/pkg/github/auth.go [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+	VerificationURI string `json:"verification_uri"`
+	ExpiresIn       int    `json:"expires_in"`
+	Interval        int    `json:"interval"`
+}
+
+// GetAccessToken requests the access token for the authorized user.
+func (df DeviceFlow) GetAccessToken(
+	req *GetAccessTokenRequest,
+) (*GetAccessTokenResponse, error) {
+	uri := df.uri("login/oauth/access_token")
```
Comment (Krlier) [inline cli/pkg/github/auth.go:55]: Would you mind adding this path to `cli/pkg/github/github.go` and use it as a constant as well? By doing this, if it changes, we'd need to simply change the value in `github.go`. Same as the `baseURL` variable.

File: cli/pkg/github/doc.go [hunk 5]
```diff
@@ -0,0 +1,2 @@
+// Package github contains utilities to work with GitHub.
```
Comment (Krlier) [inline cli/pkg/github/doc.go:1]: What is the purpose of this file? 🙂

**Review by Krlier — CHANGES_REQUESTED** [review]
Thanks for the awesome contribution, @lzakharov! Your code is looking great! 🚀

We just reviewed your code and left some comments on it. Additionally, would mind adding the project's license into all the files you've created, please?

```
// Copyright 2020 Globo.com authors. All rights reserved.
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.
```

**Comment by rafaveira3** [comment]
Thanks a lot, @lzakharov! You rock! We will review your other PR soon! 😄 🚀

File: cli/pkg/github/doc.go (hunk 5 above)
Comment (lzakharov) [reply cli/pkg/github/doc.go:5]: Just to provide general package documentation ([godoc](https://blog.golang.org/godoc)). If it's not convenient for your team, I can delete it.

Comment (Krlier) [reply cli/pkg/github/doc.go:5]: I see now, thanks! We can leave this file as it is then.

**Review by Krlier — APPROVED** [review]

**Review by rafaveira3 — APPROVED** [review]

**Comment by Krlier** [comment]
Thanks for taking the time to contribute to huskyCI, @lzakharov! Your contribution is greatly appreciated! 🚀

**Pull request merged**